File: akka/akka-wordcounter/java/src/main/java/WordCounter.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.math.BigInteger;
import java.util.StringTokenizer;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;

public class WordCounter extends AbstractActor {
	static ActorSystem actorSystem = ActorSystem.create("system");
	static long start;
	static int totalWords;
	static int totalLines;
	static int linesProcessed;

	public static void main(String[] args) {
		final ActorRef wordcounter = actorSystem.actorOf(WordCounter.props(), "wordcounter");
		wordcounter.tell(new WordCounter.Count("words.txt"), wordcounter);
	}

	public static class Count {
		String filename;

		public Count(String filename) {
			this.filename = filename;
		}
	}

	public static class Counted {
		int numWords;

		public Counted(int numWords) {
			this.numWords = numWords;
		}
	}

	public static Props props() {
		return Props.create(WordCounter.class);
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder()
			.match(Count.class, this::count)
			.match(Counted.class, this::counted)
			.build();
		}

	private void count(Count count) throws IOException {
//		System.out.println("Count(filename)");
		start = System.currentTimeMillis();
		BufferedReader br = new BufferedReader(new FileReader(count.filename));
		String line = null;
		while ((line = br.readLine()) != null) {
//			System.out.println("Line: " + line);
			totalLines++;
			ActorRef counter = actorSystem.actorOf(Counter.props());
			counter.tell(new Counter.Count(line), self());
		}
		br.close();
	}

	private void counted(Counted counted) throws IOException {
//		System.out.println("Counted " + counted.numWords + " words");
		totalWords += counted.numWords;
		linesProcessed++;
		if (linesProcessed == totalLines) {
			// All the lines have been counted
			long end = System.currentTimeMillis();
			System.out.println("" + totalWords + " in " + (end - start) + " milliseconds");
			actorSystem.terminate();
		}
	}
}

class Counter extends AbstractActor {
	public static class Count {
		String line;

		public Count(String line) {
			this.line = line;
		}
	}

	public static Props props() {
		return Props.create(Counter.class);
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder().match(Count.class, this::count).build();
	}

	private void count(Count count) {
//		System.out.println("Counting words in line: " + count.line);
		StringTokenizer st = new StringTokenizer(count.line);
		int numWords = st.countTokens();
fibonacci(100);
		sender().tell(new WordCounter.Counted(numWords), self());
	}

	// CPU intensive method to calculate the nth Fibonacci number
	static BigInteger fibonacci(long nth){
		nth = nth - 1;
		long count = 0;
		BigInteger first = BigInteger.ZERO;
		BigInteger second = BigInteger.ONE;
		BigInteger third = null;
		while (count < nth){
			third = new BigInteger(first.add(second).toString());
			first = new BigInteger(second.toString());
			second = new BigInteger(third.toString());
			count++;
		}
		return third;
	}
}

File: akka/akka-wordcounter/scala/WordCounter.scala
import scala.concurrent.duration._
import akka.actor.{ Actor, ActorLogging, ActorRef, ActorSystem, Props }
import akka.dispatch.ExecutionContexts._
import akka.pattern.ask
import akka.util.Timeout

object Main extends App {
	implicit val ec = global // TODO: What is this?
	implicit val timeout = Timeout(25 seconds)	// TODO: And this?
	val actorSystem = ActorSystem("system")
	var wordCounter = actorSystem.actorOf(Props[WordCounter], name = "wordcounter")
	//val future = wordCounter ? "words.txt" // Unrecognized message
	val future = wordCounter ? WordCounter.Count("words.txt") // The "ask"(?) method returns a Future instance
	future.map { 
		totalWords => println("Total number of words " + totalWords)
		actorSystem.terminate
	}
}

class WordCounter extends Actor {
	private var totalWords = 0
	private var totalLines = 0
	private var linesProcessed = 0
	private var fileSender: Option[ActorRef] = None

	def receive = {
		case WordCounter.Count(filename: String) => {
			fileSender = Some(sender) // Save reference to process invoker
			import scala.io.Source._
			fromFile(filename).getLines.foreach {
				// Create a separate child actor for each line and send it a message containing the line
				// NOTE: This measn that any number of lines can be counted concurrently
				line => context.actorOf(Props[Counter]) ! Counter.Count(line)
				totalLines += 1
			}
		}
		case WordCounter.Counted(numWords) => {
			totalWords += numWords
			linesProcessed += 1
			if (linesProcessed == totalLines) {
				// All the lines have been counted, so send the result back to the process invoker
				fileSender.map(_ ! totalWords)
			}
		}
		case _ => {
			println("Message not recognized!")
		}
	}
}

object WordCounter {
	case class Count(filename: String)		// Sent by main method to start counting words in the specified file
	case class Counted(numWords: Integer)	// Sent by child counters to return the number of words in a line
}

class Counter extends Actor {
	def receive = {
		case Counter.Count(string) => {
			val numWords = string.split(" ").length
			// Send the number of words in the line back to the parent actor (ie. the "sender", which is the WordCounter actor)
			sender ! WordCounter.Counted(numWords)
		}
		case _ => println("Error: message not recognized")
	}
}

object Counter {
	case class Count(string: String)
}

